File: codeforces/835/G.py
import sys
lines = list(map(str.strip, sys.stdin.readlines()))
sys.setrecursionlimit(10**6)

i = 1
while i < len(lines):
    n, a, b = map(int, lines[i].split(" "))
    adj_lists = {j: set() for j in range(1, n+1)}
    i+=1
    for _ in range(n-1):
        u, v, w = map(int, lines[i].split(" "))
        adj_lists[u].add((v, w))
        adj_lists[v].add((u, w))
        i+=1
    bsums = {}
    bseen = set()
    def dfs_b(curr, score):
        bseen.add(curr)
        for nbr, wi in adj_lists[curr]:
            if nbr in bseen: continue
            if wi^score not in bsums:
                bsums[wi^score] = [nbr]
            else:
                bsums[wi^score].append(nbr)
            dfs_b(nbr, score^wi)
    dfs_b(b, 0)
    # print(bsums)
    aseen = set()
    def dfs_a(curr, score):
        aseen.add(curr)
        if score in bsums:
            return True
        for nbr, wi in adj_lists[curr]:
            if nbr == b:
                if score^wi == 0:
                    # print("We found true when we went from", curr, "to", nbr)
                    return True
            else:
                if nbr in aseen: continue
                if score ^ wi in bsums:
                    # print("We found true when we went from", curr, "to", nbr)
                    return True
                if dfs_a(nbr, score^wi):
                    return True
        return False
    print("YES" if dfs_a(a, 0) else "NO")

File: codeforces/835/G.cpp
#include <bits/stdc++.h>

using namespace std;

void dfs_b(int curr, int score ,map<int, vector<pair<int, int>>> &adj_lists, set<int> &bsums, set<int> &bseen){
    cout << "dfs b called" << endl;
    bseen.insert(curr);
    for(auto &p: adj_lists[curr]){
        cout << p.first << " " << p.second << endl;
    }
}

int main() {
    ios::sync_with_stdio(false);
    cin.tie(0);
    int t;
    cin >> t;
    while(t--){
        int n, a, b;
        cin >> n >> a >> b;
        map<int, vector<pair<int, int>>> adj_lists;
        for(int i = 0; i < n-1; i++){
            int from, to, w;
            cin >> from >> to >> w;
            adj_lists[i+1] = {};
            adj_lists[from].push_back({to, w});
            adj_lists[to].push_back({from, w});
        }
        set<int> bsums;
        set<int> bseen;

        dfs_b(1, 0, adj_lists, bsums, bseen);
    }
    return 0;
}